### PR TITLE
DAOS-7963 Java Client build failing on Fedora

### PR DESCRIPTION
**Pull request by zjf2012**
remove the unnecessary import of "com.sun.corba.se.impl.orbutil.concurrent" which failed Java code compilation in Fedora.

Signed-off-by: jiafu zhang <jiafu.zhang@intel.com>

### CODE REVIEW

**Comment by ashleypittman** [comment]
If this is fixed then you can uncomment the lines linked below which will enable testing of this build for every PR.  It makes more sense to enable the test in the same PR as the fix rather than to test manually and have a second PR to enable the test. 

https://github.com/daos-stack/daos/blob/master/.github/workflows/ci.yml#L76-L80

**Comment by zjf2012** [comment]
> If this is fixed then you can uncomment the lines linked below which will enable testing of this build for every PR. It makes more sense to enable the test in the same PR as the fix rather than to test manually and have a second PR to enable the test.
> 
> https://github.com/daos-stack/daos/blob/master/.github/workflows/ci.yml#L76-L80

ok, it should be fixed. Let me try.

**Comment by ashleypittman** [comment]
This has failed the build, there appear to be (possibly longstanding) compiler warnings at https://github.com/daos-stack/daos/pull/6716/checks?check_run_id=3607789781#step:6:9383 which aren't fatal but look like they should be, and test failures at https://github.com/daos-stack/daos/pull/6716/checks?check_run_id=3607789781#step:6:10486 which are.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -73,11 +73,11 @@ jobs:
...
-#    - name: Build DAOS Java client in Docker image
-#      run: docker build . -f utils/docker/Dockerfile.centos.8
-#                          --build-arg COMPILER=${{ matrix.compiler }}
-#                          --build-arg DEPS_JOBS=8
-#                          --build-arg BASE_DISTRO=fedora:34
+    - name: Build DAOS Java client in Docker image
+      run: docker build . -f utils/docker/Dockerfile.centos.8
+                          --build-arg COMPILER=${{ matrix.compiler }}
+                          --build-arg DEPS_JOBS=8
+                          --build-arg BASE_DISTRO=fedora:34
```
Comment (ashleypittman) [inline .github/workflows/ci.yml:80]: FYI - you can test this code locally by running the command from here:

docker build . -f utils/docker/Dockerfile.centos.8 --build-arg BASE_DISTRO=fedora:34

If you don't have a docker server setup I can let you access one of mine on the wolf cluster.

Comment (zjf2012) [reply .github/workflows/ci.yml:80]: that would be great. please share me your docker server in wolf. thanks.

**Comment by zjf2012** [comment]
@ashleypittman finally, the issue is fixed. But I added a new dockerfile for Fodera. Please let me know if it's inappropriate.  The goal is to not install maven with JDK-11 dependencies from dnf repo. Instead, I download and install only maven by myself. And it worked with JDK-8.

**Comment by ashleypittman** [comment]
> @ashleypittman finally, the issue is fixed. But I added a new dockerfile for Fodera. Please let me know if it's inappropriate. The goal is to not install maven with JDK-11 dependencies from dnf repo. Instead, I download and install only maven by myself. And it worked with JDK-8.

Yes it is appropriate, there are other PRs doing the same thing.  What you should do however is change the dockerfile name in all parts of the ci.yaml file, and remove the BASE_DISTRO code, and change "FROM" in the fedora dockerfile.

We should try and find a way of not hardcoding version numbers in the Fedora build, is there not a repo that we can install from at all?

Happy to work with you on these two items to get it resolved.

**Comment by zjf2012** [comment]
> > @ashleypittman finally, the issue is fixed. But I added a new dockerfile for Fodera. Please let me know if it's inappropriate. The goal is to not install maven with JDK-11 dependencies from dnf repo. Instead, I download and install only maven by myself. And it worked with JDK-8.
> 
> Yes it is appropriate, there are other PRs doing the same thing. What you should do however is change the dockerfile name in all parts of the ci.yaml file, and remove the BASE_DISTRO code, and change "FROM" in the fedora dockerfile.
> 
> We should try and find a way of not hardcoding version numbers in the Fedora build, is there not a repo that we can install from at all?
> 
> Happy to work with you on these two items to get it resolved.

Just updated code accordingly. For the maven repo hard coding, I tried to get maven from Fedora repo. But maven gets installed with JDK-11 as dependency. I tried to suppress JDK-11 and configure Java to use JDK-8 with update-alternative. However, it's failed. So, I download only maven from maven site. 

I've add an arg, MAVEN_VERSION, so that we can configure maven version later.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -62,33 +62,28 @@ jobs:
       with:
         submodules: true
     - name: Prepare base image in Docker
-      run: docker build . -f utils/docker/Dockerfile.centos.8
+      run: docker build . -f utils/docker/Dockerfile.fodera
```
Comment (ashleypittman) [inline .github/workflows/ci.yml:65]: Spelling mistake.  You'll need to modify this in lots of places and rename the file.
```suggestion
      run: docker build . -f utils/docker/Dockerfile.fedora
```
Although now fedora is in it's own file you can collapse this a bit more, add ", fedora" to the matix azis on L18 and remove Lines 51 and onwards in this file.

File: utils/docker/Dockerfile.centos.8 [hunk 3]
```diff
@@ -307,6 +307,10 @@ RUN echo -e "<settings>\n\
 
 ARG DAOS_JAVA_BUILD=$DAOS_BUILD
 
+RUN CHOICE=$(echo "" | sudo alternatives --config java | grep "java-1.8" | tr -cd '[[:digit:]]') && \
+        echo ${CHOICE:0:1} | sudo alternatives --config java
+RUN java -version
+
```
Comment (ashleypittman) [inline utils/docker/Dockerfile.centos.8:313]: Should this still be here for CentOS 8?

You should remove the BASE_DISTRO arg from the top of this file however as it's no longer used.

File: utils/docker/Dockerfile.fodera [hunk 4]
```diff
@@ -0,0 +1,324 @@
+# Copyright (C) 2018-2021 Intel Corporation
+# All rights reserved.
+#
+# 'recipe' for Docker to build an image of centOS 8 based
+# environment for building the DAOS project.
+#
+
+# Pull base image
+ARG BASE_DISTRO=fedora:34
+FROM $BASE_DISTRO
```
Comment (ashleypittman) [inline utils/docker/Dockerfile.fodera:10]: Unless this is going to be modified on the command line you don't need to do this.
```suggestion
FROM fedora:34
```

File: utils/docker/Dockerfile.fodera [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+                        <mirrorOf>central</mirrorOf>\n\
+                </mirror>\n\
+        </mirrors>\n\
+</settings>" > /home/daos/.m2/settings.xml
+
+ARG DAOS_JAVA_BUILD=$DAOS_BUILD
+ARG MAVEN_VERSION=3.6.3
+USER root:root
+
+RUN [ "$DAOS_JAVA_BUILD" != "yes" ] || ( wget https://archive.apache.org/dist/maven/maven-3/$MAVEN_VERSION/binaries/apache-maven-$MAVEN_VERSION-bin.tar.gz && tar -xf apache-maven-$MAVEN_VERSION-bin.tar.gz && find ./ -name "mvn" | { read d; f=$(dirname $d); cp -r $(readlink -f $f)/../* /usr/local; } )
```
Comment (ashleypittman) [inline utils/docker/Dockerfile.fodera:313]: Is there no way at all to install the correct package, even if it would mean modifying the yum config to do it?

**Comment by johannlombardi** [comment]
What is the status of this patch? Shall we land it? Please advise.

**Comment by ashleypittman** [comment]
> What is the status of this patch? Shall we land it? Please advise.

WIP, but we should land it once complete.

**Review by brianjmurrell — DISMISSED** [review]

**Comment by ashleypittman** [comment]
This will need updating unfortunately, we have an ongoing discussion about if we spilt out the dockerfiles like this or if we keep them as one and control them with args, as you can see there's a lot of duplication if we split them and that is going to significantly increase the ongoing support burden.

Ping me on IM and we can have a look at the best way forward.

**Comment by johannlombardi** [comment]
This has been fixed already? Can we close?

**Comment by zjf2012** [comment]
> This has been fixed already? Can we close?

yes, it's fixed. But @ashleypittman seems having some concerns about the new fedora ci file. He said he'll talk to me later.

**Comment by ashleypittman** [comment]
No, it's not fixed.  Java still does not build on Fedora that I'm aware of.

**Comment by zjf2012** [comment]
> No, it's not fixed. Java still does not build on Fedora that I'm aware of.

so, I should merge my fix to your latest ci adjustment?

**Comment by ashleypittman** [comment]
> > No, it's not fixed. Java still does not build on Fedora that I'm aware of.
> 
> so, I should merge my fix to your latest ci adjustment?

Yes, but your fix is to split the centos 8 docker file into multiple files per distro which is something that we were considering but have decided to avoid if we can, so ideally if we still need different packages for different distros which are built from the same dockerfile then we'll probably need to put a conditional around the install command.

**Comment by zjf2012** [comment]
> > > No, it's not fixed. Java still does not build on Fedora that I'm aware of.
> > 
> > 
> > so, I should merge my fix to your latest ci adjustment?
> 
> Yes, but your fix is to split the centos 8 docker file into multiple files per distro which is something that we were considering but have decided to avoid if we can, so ideally if we still need different packages for different distros which are built from the same dockerfile then we'll probably need to put a conditional around the install command.

ok, I'll try to make it conditional.

File: .github/workflows/ci2.yml [hunk 6]
```diff
@@ -72,7 +72,7 @@ jobs:
           - distro: fedora
             base: centos.8
             with: fedora:35
-            java: no
+            java: yes
```
Comment (brianjmurrell) [inline .github/workflows/ci2.yml:75]: Since this ends up getting evaluated by shell script, use `true`(/`false`) here as it lets you write more language natural shell script later.

File: utils/docker/Dockerfile.centos.8 [hunk 7]
```diff
@@ -105,11 +105,20 @@ gpgcheck=False\n" >> /etc/yum.repos.d/$repo:$branch:$build_number.repo;   \
...
+ARG DAOS_BUILD=$DAOS_DEPS_BUILD
+ARG DAOS_JAVA_BUILD=$DAOS_BUILD
+ARG BASE_DISTRO
 
 # Install OS updates and package.  Include basic tools and daos dependencies
 COPY ./utils/scripts/install-centos8.sh /tmp/install.sh
 RUN chmod +x /tmp/install.sh
-RUN /tmp/install.sh
+
+RUN if [[ "$BASE_DISTRO" = fedora* ]] && [[ "$DAOS_JAVA_BUILD" == "yes" ]]; then \
```
Comment (brianjmurrell) [inline utils/docker/Dockerfile.centos.8:117]: With the previous suggestion this can become:

```suggestion
RUN if [[ "$BASE_DISTRO" = fedora* ]] && "$DAOS_JAVA_BUILD"; then \
```

File: utils/docker/Dockerfile.centos.8 [hunk 8]
```diff
@@ -105,11 +105,20 @@ gpgcheck=False\n" >> /etc/yum.repos.d/$repo:$branch:$build_number.repo;   \
...
+	. /tmp/install.sh yes; 	\
+    else 			\
+	. /tmp/install.sh no; 	\
+    fi
```
Comment (brianjmurrell) [inline utils/docker/Dockerfile.centos.8:121]: Again, use `true`/`false` for shell script booleans.

```suggestion
	. /tmp/install.sh true; 	\
    else 			\
	. /tmp/install.sh false; 	\
    fi
```

File: utils/docker/Dockerfile.centos.8 [hunk 9]
```diff
@@ -215,7 +224,7 @@ ARG COMPILER=gcc
 ARG JOBS=$DEPS_JOBS
 ARG DAOS_BUILD_TYPE=$DAOS_TARGET_TYPE
 ARG DAOS_BUILD=$DAOS_DEPS_BUILD
-
+RUN echo "DAOS_BUILD is $DAOS_BUILD"
```
Comment (brianjmurrell) [inline utils/docker/Dockerfile.centos.8:227]: Is the debug that should be removed?

File: utils/scripts/install-centos8.sh [hunk 10]
```diff
@@ -67,5 +68,20 @@ dnf install \
     python3-yaml \
     sg3_utils \
     valgrind-devel \
+    which \
     yasm
+
+echo "is fedora java? $fedora_java"
```
Comment (brianjmurrell) [inline utils/scripts/install-centos8.sh:74]: More debug that needs removing?

File: utils/scripts/install-centos8.sh [hunk 11]
```diff
@@ -67,5 +68,20 @@ dnf install \
...
+
+if [ "$fedora_java" = "yes" ]; then
```
Comment (brianjmurrell) [inline utils/scripts/install-centos8.sh:76]: Per above, use natural shell booleans:

```suggestion
if "$java"; then
```

File: utils/scripts/install-centos8.sh [hunk 12]
```diff
@@ -10,6 +10,8 @@
 
 set -e
 
+fedora_java=$1
```
Comment (brianjmurrell) [inline utils/scripts/install-centos8.sh:13]: This isn't really anything to do with _Fedora_ is it?  It's just whether to install dependencies for the Java client build, right?  Even though currently we only do that on Fedora, it could be done on other platforms also, right?

```suggestion
java=$1
```

File: utils/scripts/install-centos8.sh [hunk 13]
```diff
@@ -67,5 +68,20 @@ dnf install \
...
+
+echo "is fedora java? $fedora_java"
+
+if [ "$fedora_java" = "yes" ]; then
+	dnf install java-1.8.0-openjdk-devel
+	curl --output ./apache-maven-3.6.3-bin.tar.gz \
+https://archive.apache.org/dist/maven/maven-3/3.6.3/binaries/apache-maven-3.6.3-bin.tar.gz \
+&& tar -xf apache-maven-3.6.3-bin.tar.gz \
+&& find ./ -name "mvn" | \
+{ read -r d; f="$(dirname "$d")"; cp -r "$(readlink -f "$f")"/../* /usr/local; }
```
Comment (brianjmurrell) [inline utils/scripts/install-centos8.sh:82]: Why are we doing this for Fedora but using the `maven` RPM for other distros?

File: utils/scripts/install-centos8.sh [hunk 14]
```diff
@@ -67,5 +68,20 @@ dnf install \
...
+else
+	dnf install maven
```
Comment (brianjmurrell) [inline utils/scripts/install-centos8.sh:84]: Is `maven` even needed if the Java client is not being built?

**Review by brianjmurrell — CHANGES_REQUESTED** [review]

File: .github/workflows/ci2.yml (hunk 6 above)
Comment (zjf2012) [reply .github/workflows/ci2.yml:75]: sounds good.

File: utils/docker/Dockerfile.centos.8 (hunk 9 above)
Comment (zjf2012) [reply utils/docker/Dockerfile.centos.8:227]: removed

File: utils/scripts/install-centos8.sh (hunk 10 above)
Comment (zjf2012) [reply utils/scripts/install-centos8.sh]: removed

File: utils/scripts/install-centos8.sh (hunk 12 above)
Comment (zjf2012) [reply utils/scripts/install-centos8.sh]: yes. it's really about fedora. Java-11 is installed along with maven automatically if we install maven from repo in fedora. I tried several ways to remove the Java-11 dep but failed. Thus, I need to handle fedora separately. 

Since you mentioned no Java and maven at all if it builds without Java. I split "fedora_java" to two variables, "java" and "fedora".

File: utils/scripts/install-centos8.sh (hunk 13 above)
Comment (zjf2012) [reply utils/scripts/install-centos8.sh:86]: see above comments

File: utils/scripts/install-centos8.sh (hunk 14 above)
Comment (zjf2012) [reply utils/scripts/install-centos8.sh]: adjusted the variable. see above comments.

File: .github/workflows/ci2.yml [hunk 15]
```diff
@@ -68,15 +68,15 @@ jobs:
           - distro: rocky
             base: centos.8
             with: rockylinux/rockylinux:8
-            java: yes
+            java: true
```
Comment (ashleypittman) [inline .github/workflows/ci2.yml:71]: With the PR we should be building Java for all builds so it no longer needs to be an matrix index, you can remove these lines.

File: .github/workflows/ci2.yml [hunk 16]
```diff
@@ -110,8 +110,8 @@ jobs:
     - name: Build Java in docker.
       run: docker build . --file utils/docker/Dockerfile.${{ matrix.base }}
                           --build-arg DEPS_JOBS
-                          --build-arg BASE_DISTRO
-                          --build-arg DAOS_JAVA_BUILD
+                          --build-arg BASE_DISTRO=${{ matrix.with }}
```
Comment (ashleypittman) [inline .github/workflows/ci2.yml:113]: If no value is given here then the value from the environment will be used, which is set on L82 so the change on this line isn't required.

File: .github/workflows/ci2.yml [hunk 17]
```diff
@@ -110,8 +110,8 @@ jobs:
...
+                          --build-arg DAOS_JAVA_BUILD=${{ matrix.java }}
```
Comment (ashleypittman) [inline .github/workflows/ci2.yml:114]: You can set this to yes here directly.  It should be yes rather than true.

File: utils/docker/Dockerfile.centos.8 [hunk 18]
```diff
@@ -105,11 +105,20 @@ gpgcheck=False\n" >> /etc/yum.repos.d/$repo:$branch:$build_number.repo;   \
         dnf repolist; \
         dnf --disablerepo=\* --enablerepo=$repo:$branch:$build_number makecache; \
     done
+# For conditional build DAOS and Java
+ARG DAOS_BUILD=$DAOS_DEPS_BUILD
+ARG DAOS_JAVA_BUILD=$DAOS_BUILD
+ARG BASE_DISTRO
```
Comment (ashleypittman) [inline utils/docker/Dockerfile.centos.8:111]: Putting it this early in the dockerfile will mean that changing if java is built or not will re-install all the packages and rebuild all the dependencies, it would be better to have the customisations (or at least the DAOS_JAVA_BUILD ones) later on, preferably after L245.

**Review by ashleypittman — COMMENTED** [review]
This will need changes to landing-builds as well as ci2 to account for master builds.

If you copy the trigger changes from https://github.com/daos-stack/daos/pull/7677/files to here then landing builds will also trigger on this PR so you can verify them there.

File: utils/docker/Dockerfile.centos.8 [hunk 19]
```diff
@@ -243,7 +243,18 @@ RUN echo -e "<settings>\n\
 </settings>" > /home/daos/.m2/settings.xml
 
 ARG DAOS_JAVA_BUILD=$DAOS_BUILD
+ARG BASE_DISTRO
```
Comment (ashleypittman) [inline utils/docker/Dockerfile.centos.8:246]: You already have ARG BASE_DISTRO above, right at the start so don't need this line.

File: .github/workflows/ci2.yml [hunk 20]
```diff
@@ -68,21 +68,18 @@ jobs:
...
             base: leap.15
             with: leap.15
-            java: yes
     env:
       DEPS_JOBS: 10
       BASE_DISTRO: ${{ matrix.with }}
       DOCKER_BASE: ${{ matrix.base }}
       COMPILER: ${{ matrix.compiler }}
-      DAOS_JAVA_BUILD: ${{ matrix.java }}
+      DAOS_JAVA_BUILD: yes
```
Comment (ashleypittman) [inline .github/workflows/ci2.yml:82]: As this is set on L111 you don't need it in the env now.

File: utils/scripts/install-centos8-java.sh [hunk 21]
```diff
@@ -0,0 +1,24 @@
+#!/bin/sh
```
Comment (ashleypittman) [inline utils/scripts/install-centos8-java.sh:1]:
```suggestion
#!/bin/sh -e
```

File: utils/scripts/install-centos8.sh [hunk 22]
```diff
@@ -67,5 +65,6 @@ dnf install \
     python3-yaml \
     sg3_utils \
     valgrind-devel \
+    which \
```
Comment (ashleypittman) [inline utils/scripts/install-centos8.sh:68]: Is this change needed?

**Review by ashleypittman — COMMENTED** [review]
did you see my earlier comment about needing to update the landings triggers as well?

File: utils/docker/Dockerfile.centos.8 (hunk 19 above)
Comment (zjf2012) [reply utils/docker/Dockerfile.centos.8:246]: As tested, I cannot read any value from BASE_DISTRO if I don't redefine it again. I don't know why yet. Any idea?

Comment (ashleypittman) [reply utils/docker/Dockerfile.centos.8:246]: No, as far as I know it's just a regular arg.

File: .github/workflows/ci2.yml (hunk 20 above)
Comment (zjf2012) [reply .github/workflows/ci2.yml:82]: > did you see my earlier comment about needing to update the landings triggers as well?

I'll do it probably tomorrow.

File: utils/scripts/install-centos8.sh (hunk 22 above)
Comment (ashleypittman) [reply utils/scripts/install-centos8.sh:68]: Just checking it was intentional.  Best to leave it here and have the minimal set in the conditional script.

**Comment by ashleypittman** [comment]
Next time you push this PR can you add -ntp to the mvn commands in the dockerfiles?

```
@@ -207,7 +207,7 @@ RUN echo -e "<settings>\n\
 ARG DAOS_JAVA_BUILD=$DAOS_BUILD

 RUN [ "$DAOS_JAVA_BUILD" != "yes" ] || \
-    mvn clean install -T 1C -DskipITs -Dgpg.skip -Ddaos.install.path=/opt/daos
+    mvn clean install -ntp -T 1C -DskipITs -Dgpg.skip -Ddaos.install.path=/opt/daos
 WORKDIR /home/daos

 ARG DAOS_KEEP_SRC=no
````

**Comment by zjf2012** [comment]
> Next time you push this PR can you add -ntp to the mvn commands in the dockerfiles?
> 
> ```
> @@ -207,7 +207,7 @@ RUN echo -e "<settings>\n\
>  ARG DAOS_JAVA_BUILD=$DAOS_BUILD
> 
>  RUN [ "$DAOS_JAVA_BUILD" != "yes" ] || \
> -    mvn clean install -T 1C -DskipITs -Dgpg.skip -Ddaos.install.path=/opt/daos
> +    mvn clean install -ntp -T 1C -DskipITs -Dgpg.skip -Ddaos.install.path=/opt/daos
>  WORKDIR /home/daos
> 
>  ARG DAOS_KEEP_SRC=no
> ```

sure

File: utils/docker/Dockerfile.centos.8 (hunk 19 above)
Comment (zjf2012) [reply utils/docker/Dockerfile.centos.8]: As tested again, I need to redefine this arg to make fedora build work.

**Comment by zjf2012** [comment]
> > Next time you push this PR can you add -ntp to the mvn commands in the dockerfiles?
> > ```
> > @@ -207,7 +207,7 @@ RUN echo -e "<settings>\n\
> >  ARG DAOS_JAVA_BUILD=$DAOS_BUILD
> > 
> >  RUN [ "$DAOS_JAVA_BUILD" != "yes" ] || \
> > -    mvn clean install -T 1C -DskipITs -Dgpg.skip -Ddaos.install.path=/opt/daos
> > +    mvn clean install -ntp -T 1C -DskipITs -Dgpg.skip -Ddaos.install.path=/opt/daos
> >  WORKDIR /home/daos
> > 
> >  ARG DAOS_KEEP_SRC=no
> > ```
> 
> sure

"-ntp" is only for maven 3.6.1 and newer version. I need to use "-B -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn" in centos7 and centos8.

File: utils/scripts/install-centos8-java.sh [hunk 23]
```diff
@@ -0,0 +1,22 @@
+#!/bin/sh -e
+
+# For fedora, java-11 is installed along with maven if we install maven from
+# repo. But we need java-8 (1.8). The 'devel' package also needs to be
+# installed specifically.
+# For centos, we can install both java-8 and maven from repo.
+
+fedora=$1
+
+if "$fedora"; then
```
Comment (ashleypittman) [inline utils/scripts/install-centos8-java.sh:10]: As discussed on teams I think "dnf install maven-openjdk8" would work here, and potentially you could do the following, and have it work on any distro without reading $1 or parsing the BASE_DISTRO variable.

```dnf install maven-openjdk8 || dnf install maven```

or

```
if [ -e /etc/fedora-release]
then
    dnf install maven-openjdk8
else
   dnf install maven
fi

dnf clean all

Comment (zjf2012) [reply utils/scripts/install-centos8-java.sh:10]: sounds good. let me try.

File: utils/scripts/install-centos8-java.sh [hunk 24]
```diff
@@ -0,0 +1,14 @@
...
+# For fedora, java-11 is installed along with maven if we install maven from
+# repo. But we need java-8 (1.8). The 'devel' package also needs to be
+# installed specifically.
+# For centos, we can install both java-8 and maven from repo.
+
+if [ -e /etc/fedora-release ]; then
+	dnf install java-1.8.0-openjdk java-1.8.0-openjdk-devel maven-openjdk8
+else
+        dnf install java-1.8.0-openjdk maven
+fi
```
Comment (ashleypittman) [inline utils/scripts/install-centos8-java.sh:12]: I think at this point you can just put it in the initial install-centos8.sh, and not have the second script.
```suggestion
if [ -e /etc/fedora-release ]; then
	dnf install java-1.8.0-openjdk java-1.8.0-openjdk-devel maven-openjdk8
else
        dnf install java-1.8.0-openjdk maven
fi
```

Comment (zjf2012) [reply utils/scripts/install-centos8-java.sh:12]: It's not only for fedora, but also conditional Java build. check line 247 in dockerfile.centos.8.  

"RUN if [[ "$DAOS_JAVA_BUILD" == "yes" ]]; then "

I need to introduce a parameter to install-centos8.sh to check if it's java build.

Comment (ashleypittman) [reply utils/scripts/install-centos8-java.sh:12]: In all other cases we take the approach that we install all needed components/rpms at the beginning, regardless of if they'll be used or not.  As discussed adding additional parameters will break the docker caching of images and cause the dependencies to be rebuilt unnecessarily, so IMO we should install these packages unconditionally earlier in the process.

What you say does make sense that we should not install them until we need them, and then only if requested however we also build daos and the java client for every PR, and that means loading cached docker images from master that avoids downloading updates however with this code as it is then each PR would need to download the maven packages twice (once for gcc, once for clang) for each distro, and this introduces additional network load, but crucially an additional point of failure.

Comment (zjf2012) [reply utils/scripts/install-centos8-java.sh:12]: make sense. Let me change it.

File: utils/scripts/install-centos8.sh [hunk 25]
```diff
@@ -68,3 +67,15 @@ dnf --nodocs install \
...
+# repo. But we need java-8 (1.8). The 'devel' package also needs to be
+# installed specifically.
+
+if [ -e /etc/fedora-release ]; then
+        dnf install java-1.8.0-openjdk-devel maven-openjdk8
+else
+        dnf install maven
+fi
+
+dnf clean all
```
Comment (ashleypittman) [inline utils/scripts/install-centos8.sh:81]: This line isn't needed now, the dockerfile will do this after calling the scipt.

Comment (zjf2012) [reply utils/scripts/install-centos8.sh]: removed

**Review by ashleypittman — APPROVED** [review]

**Review by johannlombardi — APPROVED** [review]
